--- monero/src/rpc/message_data_structs.h	2020-08-05 16:35:28.000000000 +0200
+++ monero-new/src/rpc/message_data_structs.h	2020-08-05 16:32:49.000000000 +0200
@@ -33,6 +33,7 @@
 #include "cryptonote_basic/difficulty.h"
 #include "ringct/rctSigs.h"
 #include "rpc/rpc_handler.h"
+#include "offshore/pricing_record.h"
 
 #include <unordered_map>
 #include <vector>
@@ -163,6 +164,7 @@ namespace rpc
     uint64_t timestamp;
     crypto::hash  prev_id;
     uint32_t nonce;
+    offshore::pricing_record pricing_record;
     uint64_t height;
     uint64_t depth;
     crypto::hash hash;
