diff --git a/src/rpc/rpc_version_str.cpp b/src/rpc/rpc_version_str.cpp
index c60cf4891..f8b0cf203 100644
--- a/src/rpc/rpc_version_str.cpp
+++ b/src/rpc/rpc_version_str.cpp
@@ -36,6 +36,7 @@ namespace cryptonote
 namespace rpc
 {
 
+// The following comment isn't correct for haven
 // Expected format of Monero software version string:
 // 1) Four numbers, one to two digits each, separated by periods
 // 2) Optionally, one of the following suffixes:
@@ -45,7 +46,7 @@ namespace rpc
 bool is_version_string_valid(const std::string& str)
 {
     return std::regex_match(str, std::regex(
-        "^\\d{1,2}(\\.\\d{1,2}){3}(-(release|[0-9a-f]{9}))?$",
+        "^(\\d{1,2}\\.){2}\\d{1,2}[a-z]-[a-z0-9]*, based on Monero (\\d{1,2}\\.){3}\\d{1,2}-[a-z0-9]*$",
         std::regex_constants::nosubs
     ));
 }
