--- monero/src/blockchain_utilities/blockchain_export.cpp	2020-08-05 16:35:27.000000000 +0200
+++ monero-new/src/blockchain_utilities/blockchain_export.cpp	2020-08-05 16:32:48.000000000 +0200
@@ -97,7 +97,7 @@ int main(int argc, char* argv[])
 
   if (command_line::get_arg(vm, command_line::arg_help))
   {
-    std::cout << "Monero '" << MONERO_RELEASE_NAME << "' (v" << MONERO_VERSION_FULL << ")" << ENDL << ENDL;
+    std::cout << "Haven '" << MONERO_RELEASE_NAME << "' (v" << MONERO_VERSION_FULL << ")" << ENDL << ENDL;
     std::cout << desc_options << std::endl;
     return 1;
   }
