--- monero/src/wallet/CMakeLists.txt	2020-08-05 16:35:28.000000000 +0200
+++ monero-new/src/wallet/CMakeLists.txt	2020-08-05 16:32:48.000000000 +0200
@@ -107,7 +107,7 @@ target_link_libraries(wallet_rpc_server
     ${EXTRA_LIBRARIES})
 set_property(TARGET wallet_rpc_server
   PROPERTY
-    OUTPUT_NAME "monero-wallet-rpc")
+    OUTPUT_NAME "haven-wallet-rpc")
 install(TARGETS wallet_rpc_server DESTINATION bin)
 
 
