diff --git a/src/wallet/wallet_rpc_payments.cpp b/src/wallet/wallet_rpc_payments.cpp
index 4f5364269..5709789ad 100644
--- a/src/wallet/wallet_rpc_payments.cpp
+++ b/src/wallet/wallet_rpc_payments.cpp
@@ -151,7 +151,8 @@ bool wallet2::search_for_rpc_payment(uint64_t credits_target, const std::functio
     else
     {
       int cn_variant = hashing_blob[0] >= 7 ? hashing_blob[0] - 6 : 0;
-      crypto::cn_slow_hash(hashing_blob.data(), hashing_blob.size(), hash, cn_variant, height);
+      cn_pow_hash_v3 cph;
+      cph.hash(hashing_blob.data(), hashing_blob.size(), hash.data);
     }
     ++n_hashes;
     if (cryptonote::check_hash(hash, diff))
