--- monero/src/common/compat/glibc_compat.cpp	2020-08-05 16:35:27.000000000 +0200
+++ monero-new/src/common/compat/glibc_compat.cpp	2020-08-05 16:32:48.000000000 +0200
@@ -82,7 +82,7 @@ __explicit_bzero_chk (void *dst, size_t 
 #undef glob
 extern "C" int glob_old(const char * pattern, int flags, int (*errfunc) (const char *epath, int eerrno), glob_t *pglob);
 #ifdef __i386__
-__asm__(".symver glob_old,glob@GLIBC_2.0");
+__asm__(".symver glob_old,glob@GLIBC_2.1");
 #elif defined(__amd64__)
 __asm__(".symver glob_old,glob@GLIBC_2.2.5");
 #elif defined(__arm__)
