--- src/blockchain_db/blockchain_db.h.orig	2020-08-14 16:26:32.878148600 +0200
+++ src/blockchain_db/blockchain_db.h	2020-08-14 16:46:41.275437000 +0200
@@ -166,12 +166,15 @@
   uint8_t relayed;
   uint8_t do_not_relay;
   uint8_t double_spend_seen: 1;
+  uint64_t fee_usd;
+  uint64_t offshore_fee;
+  uint64_t offshore_fee_usd;
   uint8_t pruned: 1;
   uint8_t is_local: 1;
   uint8_t dandelionpp_stem : 1;
   uint8_t bf_padding: 4;
 
-  uint8_t padding[76]; // till 192 bytes
+  uint8_t padding[52]; // till 192 bytes
 
   void set_relay_method(relay_method method) noexcept;
   relay_method get_relay_method() const noexcept;
@@ -1151,6 +1154,13 @@
    */
   virtual uint64_t height() const = 0;
 
+  /**
+   * @brief fetch the circulating supply tally values from the blockchain
+   *
+   * @return the current circulating supply tally values
+   */
+  virtual std::vector<std::pair<std::string, int64_t>> get_circulating_supply() const = 0;
+  
 
   /**
    * <!--
