--- monero/src/debug_utilities/cn_deserialize.cpp	2019-12-20 13:52:25.292151421 +0000
+++ monero-offshore/src/debug_utilities/cn_deserialize.cpp	2019-11-27 08:35:55.856825253 +0000
@@ -103,7 +103,7 @@ int main(int argc, char* argv[])
 
   if (command_line::get_arg(vm, command_line::arg_help))
   {
-    std::cout << "Monero '" << MONERO_RELEASE_NAME << "' (v" << MONERO_VERSION_FULL << ")" << ENDL << ENDL;
+    std::cout << "Haven '" << MONERO_RELEASE_NAME << "' (v" << MONERO_VERSION_FULL << ")" << ENDL << ENDL;
     std::cout << desc_options << std::endl;
     return 1;
   }
