--- cryptonote_protocol/cryptonote_protocol_handler.inl.orig	2020-08-14 16:26:34.262298700 +0200
+++ cryptonote_protocol/cryptonote_protocol_handler.inl	2020-08-14 19:07:13.701992100 +0200
@@ -2335,7 +2335,7 @@
         }
       }
       MGINFO_YELLOW(ENDL << "**********************************************************************" << ENDL
-        << "You are now synchronized with the network. You may now start monero-wallet-cli." << ENDL
+        << "You are now synchronized with the network. You may now start haven-wallet-cli." << ENDL
         << ENDL
         << "Use the \"help\" command to see the list of available commands." << ENDL
         << "**********************************************************************");
@@ -2657,7 +2657,7 @@
       m_core.set_target_blockchain_height(target);
       if (target == 0 && context.m_state > cryptonote_connection_context::state_before_handshake && !m_stopping)
       {
-        MCWARNING("global", "monerod is now disconnected from the network");
+        MCWARNING("global", "havend is now disconnected from the network");
         m_ask_for_txpool_complement = true;
       }
     }
