--- monero/src/gen_multisig/CMakeLists.txt	2020-08-05 16:35:28.000000000 +0200
+++ monero-new/src/gen_multisig/CMakeLists.txt	2020-08-05 16:32:48.000000000 +0200
@@ -50,5 +50,5 @@ add_dependencies(gen_multisig
   version)
 set_property(TARGET gen_multisig
   PROPERTY
-    OUTPUT_NAME "monero-gen-trusted-multisig")
+    OUTPUT_NAME "haven-gen-trusted-multisig")
 install(TARGETS gen_multisig DESTINATION bin)
