diff --git a/src/serialization/string.h b/src/serialization/string.h
index fd8450f7b..bebb323f5 100644
--- a/src/serialization/string.h
+++ b/src/serialization/string.h
@@ -31,6 +31,7 @@
 #pragma once
 #include <memory>
 #include "serialization.h"
+#include "json_archive.h"
 
 template <template <bool> class Archive>
 inline bool do_serialize(Archive<false>& ar, std::string& str)
@@ -54,8 +55,13 @@ inline bool do_serialize(Archive<false>& ar, std::string& str)
 template <template <bool> class Archive>
 inline bool do_serialize(Archive<true>& ar, std::string& str)
 {
-  size_t size = str.size();
-  ar.serialize_varint(size);
-  ar.serialize_blob(const_cast<std::string::value_type*>(str.c_str()), size);
+  if (typeid(ar) == typeid(json_archive<true>)) {
+    size_t size = str.size();
+    ar.serialize_readable_string(const_cast<std::string::value_type*>(str.c_str()), size);
+  } else {
+    size_t size = str.size();
+    ar.serialize_varint(size);
+    ar.serialize_blob(const_cast<std::string::value_type*>(str.c_str()), size);
+  }
   return true;
 }
