diff --git a/src/cryptonote_core/tx_sanity_check.h b/src/cryptonote_core/tx_sanity_check.h
index 4a469462f..d6a5cf3a1 100644
--- a/src/cryptonote_core/tx_sanity_check.h
+++ b/src/cryptonote_core/tx_sanity_check.h
@@ -31,6 +31,6 @@
 
 namespace cryptonote
 {
-  bool tx_sanity_check(const cryptonote::blobdata &tx_blob, uint64_t rct_outs_available);
+  bool tx_sanity_check(const cryptonote::blobdata &tx_blob, uint64_t rct_outs_available, uint64_t current_height);
   bool tx_sanity_check(const std::set<uint64_t> &rct_indices, size_t n_indices, uint64_t rct_outs_available);
 }
