--- monero/src/cryptonote_core/tx_sanity_check.cpp	2020-08-04 19:07:58.000000000 +0200
+++ monero-new/src/cryptonote_core/tx_sanity_check.cpp	2020-08-03 21:31:15.000000000 +0200
@@ -54,20 +54,48 @@ bool tx_sanity_check(Blockchain &blockch
     MERROR("Transaction is coinbase");
     return false;
   }
+
+  if ((tx.version >= OFFSHORE_TRANSACTION_VERSION) && (tx.pricing_record_height > 0)) {
+    // Validate that pricing record is not too old
+    uint64_t current_height = blockchain.get_current_blockchain_height();
+    if ((current_height - PRICING_RECORD_VALID_BLOCKS) > tx.pricing_record_height) {
+      MERROR("Offshore transaction references a pricing record that is too old - rejected");
+      return false;
+    }
+  }
+  
   std::set<uint64_t> rct_indices;
   size_t n_indices = 0;
 
   for (const auto &txin : tx.vin)
   {
-    if (txin.type() != typeid(cryptonote::txin_to_key))
-      continue;
-    const cryptonote::txin_to_key &in_to_key = boost::get<cryptonote::txin_to_key>(txin);
-    if (in_to_key.amount != 0)
+    if (txin.type() == typeid(cryptonote::txin_to_key)) {
+      const cryptonote::txin_to_key &in_to_key = boost::get<cryptonote::txin_to_key>(txin);
+      if (in_to_key.amount != 0)
+	continue;
+      const std::vector<uint64_t> absolute = cryptonote::relative_output_offsets_to_absolute(in_to_key.key_offsets);
+      for (uint64_t offset: absolute)
+	rct_indices.insert(offset);
+      n_indices += in_to_key.key_offsets.size();
+    } else if (txin.type() == typeid(cryptonote::txin_offshore)) {
+      const cryptonote::txin_offshore &in_to_key = boost::get<cryptonote::txin_offshore>(txin);
+      if (in_to_key.amount != 0)
+	continue;
+      const std::vector<uint64_t> absolute = cryptonote::relative_output_offsets_to_absolute(in_to_key.key_offsets);
+      for (uint64_t offset: absolute)
+	rct_indices.insert(offset);
+      n_indices += in_to_key.key_offsets.size();
+    } else if (txin.type() == typeid(cryptonote::txin_onshore)) {
+      const cryptonote::txin_onshore &in_to_key = boost::get<cryptonote::txin_onshore>(txin);
+      if (in_to_key.amount != 0)
+	continue;
+      const std::vector<uint64_t> absolute = cryptonote::relative_output_offsets_to_absolute(in_to_key.key_offsets);
+      for (uint64_t offset: absolute)
+	rct_indices.insert(offset);
+      n_indices += in_to_key.key_offsets.size();
+    } else {
       continue;
-    const std::vector<uint64_t> absolute = cryptonote::relative_output_offsets_to_absolute(in_to_key.key_offsets);
-    for (uint64_t offset: absolute)
-      rct_indices.insert(offset);
-    n_indices += in_to_key.key_offsets.size();
+    }
   }
 
   if (n_indices <= 10)
