--- monero/src/daemon/command_line_args.h	2019-12-20 13:52:25.288151367 +0000
+++ monero-offshore/src/daemon/command_line_args.h	2019-11-27 08:35:56.004826801 +0000
@@ -35,7 +35,7 @@
 
 namespace daemon_args
 {
-  std::string const WINDOWS_SERVICE_NAME = "Monero Daemon";
+  std::string const WINDOWS_SERVICE_NAME = "Haven Daemon";
 
   const command_line::arg_descriptor<std::string, false, true, 2> arg_config_file = {
     "config-file"
