--- monero/src/daemon/command_server.cpp	2020-08-05 16:35:28.000000000 +0200
+++ monero-new/src/daemon/command_server.cpp	2020-08-05 16:32:48.000000000 +0200
@@ -363,7 +363,7 @@ bool t_command_server::help(const std::v
 std::string t_command_server::get_commands_str()
 {
   std::stringstream ss;
-  ss << "Monero '" << MONERO_RELEASE_NAME << "' (v" << MONERO_VERSION_FULL << ")" << std::endl;
+  ss << "Haven '" << MONERO_RELEASE_NAME << "' (v" << MONERO_VERSION_FULL << ")" << std::endl;
   ss << "Commands: " << std::endl;
   std::string usage = m_command_lookup.get_usage();
   boost::replace_all(usage, "\n", "\n  ");
