--- monero/src/daemon/rpc_command_executor.cpp	2020-08-04 19:07:58.000000000 +0200
+++ monero-new/src/daemon/rpc_command_executor.cpp	2020-08-03 21:31:15.000000000 +0200
@@ -1344,10 +1344,10 @@ bool t_rpc_command_executor::print_statu
   bool daemon_is_alive = m_rpc_client->check_connection();
 
   if(daemon_is_alive) {
-    tools::success_msg_writer() << "monerod is running";
+    tools::success_msg_writer() << "havend is running";
   }
   else {
-    tools::fail_msg_writer() << "monerod is NOT running";
+    tools::fail_msg_writer() << "havend is NOT running";
   }
 
   return true;
