--- monero/src/cryptonote_basic/miner.cpp	2020-08-04 19:07:58.000000000 +0200
+++ monero-new/src/cryptonote_basic/miner.cpp	2020-08-03 21:31:15.000000000 +0200
@@ -81,8 +81,6 @@ using namespace epee;
 #include "miner.h"
 
 
-extern "C" void slow_hash_allocate_state();
-extern "C" void slow_hash_free_state();
 namespace cryptonote
 {
 
@@ -523,7 +521,6 @@ namespace cryptonote
     difficulty_type local_diff = 0;
     uint32_t local_template_ver = 0;
     block b;
-    slow_hash_allocate_state();
     ++m_threads_active;
     while(!m_stop)
     {
@@ -588,7 +585,6 @@ namespace cryptonote
       ++m_hashes;
       ++m_total_hashes;
     }
-    slow_hash_free_state();
     MGINFO("Miner thread stopped ["<< th_local_index << "]");
     --m_threads_active;
     return true;
