--- monero/src/ringct/multiexp.cc	2020-08-04 19:07:58.000000000 +0200
+++ monero-new/src/ringct/multiexp.cc	2020-08-03 21:31:15.000000000 +0200
@@ -447,7 +447,6 @@ rct::key straus(const std::vector<Multie
 {
   CHECK_AND_ASSERT_THROW_MES(cache == NULL || cache->size >= data.size(), "Cache is too small");
   MULTIEXP_PERF(PERF_TIMER_UNIT(straus, 1000000));
-  bool HiGi = cache != NULL;
   STEP = STEP ? STEP : 192;
 
   MULTIEXP_PERF(PERF_TIMER_START_UNIT(setup, 1000000));
