--- ringct/rctTypes.cpp.orig	2020-08-14 16:19:03.623859000 +0200
+++ ringct/rctTypes.cpp	2020-08-14 19:37:03.850927000 +0200
@@ -195,6 +195,7 @@
             case RCTTypeSimple:
             case RCTTypeBulletproof:
             case RCTTypeBulletproof2:
+            case RCTTypeCLSAG:
                 return true;
             default:
                 return false;
@@ -207,6 +208,7 @@
         {
             case RCTTypeBulletproof:
             case RCTTypeBulletproof2:
+            case RCTTypeCLSAG:
                 return true;
             default:
                 return false;
