diff --git a/src/blockchain_db/lmdb/db_lmdb.h b/src/blockchain_db/lmdb/db_lmdb.h
index 4b46f081..e38a98aa 100644
--- a/src/blockchain_db/lmdb/db_lmdb.h
+++ b/src/blockchain_db/lmdb/db_lmdb.h
@@ -70,6 +70,12 @@ typedef struct mdb_txn_cursors
   MDB_cursor *m_txc_hf_versions;
 
   MDB_cursor *m_txc_properties;
+
+  // HERE BE DRAGONS!!!
+  // NEAC : Add cursor for the circulating supply data
+  MDB_cursor *m_txc_circ_supply;
+  MDB_cursor *m_txc_circ_supply_tally;
+  // LAND AHOY!!!
 } mdb_txn_cursors;
 
 #define m_cur_blocks	m_cursors->m_txc_blocks
@@ -89,6 +95,8 @@ typedef struct mdb_txn_cursors
 #define m_cur_txpool_blob	m_cursors->m_txc_txpool_blob
 #define m_cur_hf_versions	m_cursors->m_txc_hf_versions
 #define m_cur_properties	m_cursors->m_txc_properties
+#define m_cur_circ_supply       m_cursors->m_txc_circ_supply
+#define m_cur_circ_supply_tally m_cursors->m_txc_circ_supply_tally
 
 typedef struct mdb_rflags
 {
@@ -110,6 +118,8 @@ typedef struct mdb_rflags
   bool m_rf_txpool_blob;
   bool m_rf_hf_versions;
   bool m_rf_properties;
+  bool m_rf_circ_supply;
+  bool m_rf_circ_supply_tally;
 } mdb_rflags;
 
 typedef struct mdb_threadinfo
@@ -241,6 +251,8 @@ public:
 
   virtual block get_top_block() const;
 
+  virtual std::vector<std::pair<std::string, int64_t>> get_circulating_supply() const;
+  
   virtual uint64_t height() const;
 
   virtual bool tx_exists(const crypto::hash& h) const;
@@ -456,7 +468,10 @@ private:
   MDB_dbi m_hf_versions;
 
   MDB_dbi m_properties;
-
+  
+  MDB_dbi m_circ_supply;
+  MDB_dbi m_circ_supply_tally;
+  
   mutable uint64_t m_cum_size;	// used in batch size estimation
   mutable unsigned int m_cum_count;
   std::string m_folder;
